### PR TITLE
Remove the legacy wallet and BDB dependency

### PR DESCRIPTION
**Pull request by achow101**
The final step of #20160.

A bare minimum of legacy wallet code is kept in order to perform wallet migration. Migration of legacy wallets uses the independent BDB parser and a minimal `LegacyDataSPKM` that allows the legacy data to be loaded so that the migration can be completed.

BDB has been removed as a dependency and documentation have been updated to reflect that.

### CODE REVIEW

**Review by jess2505 — APPROVED** [review]

File: src/wallet/migrate.cpp [hunk 1]
```diff
@@ -1,13 +1,601 @@
...
+
+    PAGETYPE_MAX = 14,
+};
+
+enum class RecordType : uint8_t
+{
+    KEYDATA = 1,
+    DUPLICATE = 2,
+    OVERFLOW_DATA = 3,
+    DELETE = 0x80, // Indicate this record is deleted. This is AND'd with the real type.
```
Comment (laanwj) [inline src/wallet/migrate.cpp:47]: OR'ed, i guess?

**Comment by Sjors** [comment]
You can also drop the `BerkeleyDatabaseSanityCheck` related suppression in `contrib/devtools/check-devs.sh`

**Comment by achow101** [comment]
> It looks like all three dependencies got merged, is this ready for review?

Currently it is still dependent on #30328 but I suppose it doesn't have to be.

File: src/wallet/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -79,7 +81,7 @@ bool HaveKeys(const std::vector<valtype>& pubkeys, const LegacyDataSPKM& keystor
 //!                            scripts or simply treat any script that has been
 //!                            stored in the keystore as spendable
 // NOLINTNEXTLINE(misc-no-recursion)
-IsMineResult IsMineInner(const LegacyDataSPKM& keystore, const CScript& scriptPubKey, IsMineSigVersion sigversion, bool recurse_scripthash=true)
+IsMineResult LegacyWalletIsMineInnerDONOTUSE(const LegacyDataSPKM& keystore, const CScript& scriptPubKey, IsMineSigVersion sigversion, bool recurse_scripthash=true)
```
Comment (Sjors) [inline src/wallet/scriptpubkeyman.cpp:84]: e6ccd17275e06e6278473587257f80c2d9bd8834: maybe move the legacy wallet migration code to a separate file? And mark it read-only :-)

**Comment by onlinesipahimithu** [comment]
Helli how to develop

**Comment by achow101** [comment]
The test each commit task fails because the HEAD of this PR has bdb removed from that task's install list, but the earlier commits in this PR still needs BDB. I guess we can leave it there for now and remove it later?

File: doc/files.md [hunk 3]
```diff
@@ -123,6 +114,15 @@ Path           | Description | Repository notes
 `addr.dat`     | Peer IP address BDB database; replaced by `peers.dat` in [0.7.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.7.0.md) | [PR #1198](https://github.com/bitcoin/bitcoin/pull/1198), [`928d3a01`](https://github.com/bitcoin/bitcoin/commit/928d3a011cc66c7f907c4d053f674ea77dc611cc)
 `onion_private_key` | Cached Tor onion service private key for `-listenonion` option. Was used for Tor v2 services; replaced by `onion_v3_private_key` in [0.21.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.21.0.md) | [PR #19954](https://github.com/bitcoin/bitcoin/pull/19954)
 
+### Berkeley DB database based wallets
+
```
[the remark this answers is not in the document]
Comment (achow101) [reply doc/files.md:118]: I think it's fine to leave as is, it's under the "Legacy subdirectories and Files" heading. This document discusses the files that Bitcoin Core creates and uses, and legacy wallets are neither created nor really used after this PR.

File: CMakePresets.json [hunk 4]
```diff
@@ -80,7 +80,6 @@
         "ENABLE_HARDENING": "ON",
         "ENABLE_WALLET": "ON",
         "WARN_INCOMPATIBLE_BDB": "OFF",
```
[the remark this answers is not in the document]
Comment (achow101) [reply CMakePresets.json:82]: Done.

**Comment by achow101** [comment]
Rebased and added a couple commits for followups from #31250

File: depends/packages/bdb.mk [hunk 5]
```diff
@@ -1,33 +0,0 @@
...
-$(package)_config_opts=--disable-shared --enable-cxx --disable-replication --enable-option-checking
-$(package)_config_opts_mingw32=--enable-mingw
-$(package)_cflags+=-Wno-error=implicit-function-declaration -Wno-error=format-security -Wno-error=implicit-int
-$(package)_cppflags_freebsd=-D_XOPEN_SOURCE=600 -D__BSD_VISIBLE=1
-$(package)_cppflags_netbsd=-D_XOPEN_SOURCE=600
-$(package)_cppflags_mingw32=-DUNICODE -D_UNICODE
-endef
-
-define $(package)_preprocess_cmds
-  patch -p1 < $($(package)_patch_dir)/clang_cxx_11.patch && \
```
Comment (fanquake) [inline depends/packages/bdb.mk:19]: `clang_cxx_11.patch` also needs deleting.

Comment (achow101) [reply depends/packages/bdb.mk:19]: Removed

**Comment by achow101** [comment]
> There's also a `libdb++-dev` instance in `workflows/ci.yml` & `deadlock:libdb` / `BerkeleyBatch` etc in tsan suppressions. `Berkeley*` in walletdb.h. `BerkeleyEnvironment::Salvage` in `utils_tests.cpp`.

Removed

File: src/wallet/interfaces.cpp [hunk 6]
```diff
@@ -177,11 +177,7 @@ class WalletImpl : public Wallet
     }
     bool haveWatchOnly() override
     {
-        auto spk_man = m_wallet->GetLegacyScriptPubKeyMan();
-        if (spk_man) {
-            return spk_man->HaveWatchOnly();
-        }
-        return false;
+        return m_wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS);
```
Comment (Sjors) [inline src/wallet/interfaces.cpp:180]: In dc7bf5fd6a320c4528a28cef2a565366bbab3877 "wallet: Delete LegacySPKM": why doesn't this just return `false` like before? And if so, can't the entire helper be dropped?

File: src/wallet/test/CMakeLists.txt [hunk 7]
```diff
@@ -13,9 +13,7 @@ target_sources(test_bitcoin
     feebumper_tests.cpp
     group_outputs_tests.cpp
     init_tests.cpp
-    ismine_tests.cpp
```
Comment (Sjors) [inline src/wallet/test/CMakeLists.txt:16]: In dc7bf5fd6a320c4528a28cef2a565366bbab3877 "wallet: Delete LegacySPKM": this test (file) and the one below still exist.

**Review by Sjors — CHANGES_REQUESTED** [review]
Concept ACK.

bc3f07e384c2e145d6d14cfa3ad65b976233b538 looks good, except:

1. The GUI now shows a watch-only balance for descriptor wallets. See inline for the culprit.

2. In that same commit, the churn in `wallet/scriptpubkeyman.cpp` is hard to follow, both here on Github as well as a `git show --color-moved=dimmed-zebra`. I'm guessing you're moving stuff around, deleting and modifying? _Update_: use `--histogram`, it's just deletes.

And a few details:
- there's one more `-DWITH_BDB=ON` in `test-each-commit-exec.sh`
- `test_framework/bdb.py` seems unused now
- `fuzz/wallet_dbd_parser` checks for the presence of `USE_BDB` which is dropped.
- `DEFAULT_FLUSHWALLET` is unused
- `wallet_createwallet.py` has a comment referring to `sethdseed`
- `importprivkey` and friends are still in `bitcoin-cli.bash`

Also it might be good to rename the `importprivkey` test helper to e.g. `importkey`, so it's not confused with the disappeared RPC method.

As I mentioned in an earlier review https://github.com/bitcoin/bitcoin/pull/28710/commits/bc3f07e384c2e145d6d14cfa3ad65b976233b538#r1913267051, rather than naming functions `DONOTUSE` you could also move the migration code into a new file. But that could be a nice followup, along with changes to remove any legacy wallet weirdness from `scriptpubkeyman.h`

**Review by maflcko — APPROVED** [review]
lgtm ACK bc3f07e384c2e145d6d14cfa3ad65b976233b538 💇

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK bc3f07e384c2e145d6d14cfa3ad65b976233b538 💇
eF2eywtBy1XVEkBKNQpSjNzCFy6rQijB2u0oSJ5ekqHuSEbmLrtrlwP1Y1QQ9QHaog9QUeesH7zr4jfYiCYqAg==
```

</details>

File: src/wallet/interfaces.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply src/wallet/interfaces.cpp:180]: It causes a bug in the GUI, see (1) here: https://github.com/bitcoin/bitcoin/pull/28710#pullrequestreview-2798630677

**Comment by Sjors** [comment]
@maflcko `--patience` did the trick. It can be made default too if your computer is fast enough: `git config --global diff.algorithm patience`

File: src/qt/overviewpage.cpp
[the remark this answers is not in the document]
Comment (achow101) [reply src/qt/overviewpage.cpp:1]: Probably, but I'm going to leave that for a followup, this diff is large enough already.

File: src/wallet/walletdb.cpp [hunk 8]
```diff
@@ -744,18 +729,6 @@ static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch,
...
 
-    // Load keypool
-    LoadResult pool_res = LoadRecords(pwallet, batch, DBKeys::POOL,
-        [] (CWallet* pwallet, DataStream& key, DataStream& value, std::string& err) {
-        int64_t nIndex;
-        key >> nIndex;
-        CKeyPool keypool;
-        value >> keypool;
-        pwallet->GetOrCreateLegacyDataSPKM()->LoadKeyPool(nIndex, keypool);
-        return DBErrors::LOAD_OK;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/walletdb.cpp:758]: Descriptor wallets can't use non-HD keys in the "keypool", so it doesn't make sense to keep this for such wallets.

All migrated wallets get new descriptors anyways for new addresses. The keypool is entirely dumped.

File: src/wallet/interfaces.cpp [hunk 9]
```diff
@@ -587,7 +587,7 @@ class WalletLoaderImpl : public WalletLoader
         m_context.scheduler = &scheduler;
         return StartWallets(m_context);
     }
-    void flush() override { return FlushWallets(m_context); }
+    void flush() override {}
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/interfaces.cpp:590]: `flush()` is inherited from `ChainClient`, which defines it as pure virtual, so it needs to have an implementation here even if it does nothing.

File: src/wallet/interfaces.cpp (hunk 6 above)
Comment (achow101) [reply src/wallet/interfaces.cpp]: Indeed, removed.

File: src/wallet/test/CMakeLists.txt (hunk 7 above)
Comment (achow101) [reply src/wallet/test/CMakeLists.txt]: Apparently since this PR was opened, descriptor specific tests were added to those files. I've removed the legacy ones from them and added back to the `CMakeLists.txt`.

File: src/wallet/interfaces.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/interfaces.cpp]: Removed

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -2776,68 +2667,6 @@ void CWallet::ListLockedCoins(std::vector<COutPoint>& vOutpts) const
     }
 }
 
-/** @} */ // end of Actions
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp:2779]: Done

File: src/wallet/wallet.cpp [hunk 11]
```diff
@@ -235,8 +235,7 @@ static std::set<std::string> g_unloading_wallet_set GUARDED_BY(g_wallet_release_
 static void FlushAndDeleteWallet(CWallet* wallet)
 {
     const std::string name = wallet->GetName();
-    wallet->WalletLogPrintf("Releasing wallet %s..\n", name);
-    wallet->Flush();
+    wallet->WalletLogPrintf("Releasing wallet\n");
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp]: Removed

**Comment by achow101** [comment]
> And a few details:

Removed these

> Also it might be good to rename the `importprivkey` test helper to e.g. `importkey`, so it's not confused with the disappeared RPC method.

I'm going to leave that for a followup.

> You can use `--patience`. Maybe add this to the commit message?

Done, also mentioned `--histogram`.

**Comment by Sjors** [comment]
The `updateWatchOnlyLabels` stuff is still wrong. I think the entire second column for watch-only needs to be stripped from the GUI. But at least not rendered.

**Comment by achow101** [comment]
> The `updateWatchOnlyLabels` stuff is still wrong. I think the entire second column for watch-only needs to be stripped from the GUI. But at least not rendered.

Fixed it so it doesn't render. The full removal can be done in a followup.

**Comment by Sjors** [comment]
ACK 1c0d89358e12fc871e99c8304d5cb50965cf7b9d

🌲 🪚

**Review by Sjors — APPROVED** [review]
ACK 1c0d89358e12fc871e99c8304d5cb50965cf7b9d 🌲 🪚

**Review by furszy — COMMENTED** [review]
In 425c39c6f520a1b0d18c6e782141f711509b2947:
Could probably also remove the proxy functions we have in the test framework for those RPC commands. Look [at this line](https://github.com/bitcoin/bitcoin/blob/fc6346dbc8dc3db40aad4079210332b5f8b332ed/test/functional/test_framework/test_node.py#L979) for example (same for importpubkey, importprivkey and addmultisigaddress).

Also, on an orthogonal topic — I'm wondering if, instead of completely removing `importaddress()`, we might want to quickly make it compatible with descriptor wallets (e.g., by reviving #27034). This is also connected to #30175.
I’m bringing this topic up because removing something in one release and reintroducing it in another one could be annoying for users.

**Review by maflcko — APPROVED** [review]
re-review ACK 1c0d89358e 🎽

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-review ACK 1c0d89358e 🎽
PZKgfTbviYeE2sXUccefHEGzNEiLnxq231kR9x+Kbs/03JOhEOQNRuGw8xDKO5Oz1bBAAIPB1ljq05mIAK6UBQ==
```

</details>


Only stuff is some follow-up ideas:

* https://github.com/bitcoin/bitcoin/pull/28710#issuecomment-2842132320
* https://github.com/bitcoin/bitcoin/pull/28710#discussion_r2063317183
* https://github.com/bitcoin/bitcoin/pull/28710#discussion_r2063011962
* https://github.com/bitcoin/bitcoin/pull/28710#discussion_r2063234336

File: test/functional/wallet_createwallet.py [hunk 12]
```diff
@@ -27,7 +26,7 @@ def skip_test_if_missing_module(self):
 
     def run_test(self):
         node = self.nodes[0]
-        self.generate(node, 1) # Leave IBD for sethdseed
+        self.generate(node, 1)
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/wallet_createwallet.py]: Done both.

File: doc/files.md [hunk 13]
```diff
@@ -8,14 +8,14 @@
```
[the remark this answers is not in the document]
Comment (achow101) [reply doc/files.md:8]: Removed:
* `importmulti' in doc/descriptors.md
* bdb.cpp in linter
* `WriteHDCHain` and WriteCScript`

I would prefer to leave further cleanups for follow up RPs as I am sure there are plenty of things that were missed and trying to catch them all in this one PR is going to be very annoying and just delay this even more.

File: src/wallet/wallettool.cpp [hunk 14]
```diff
@@ -34,8 +34,7 @@ static void WalletCreate(CWallet* wallet_instance, uint64_t wallet_creation_flag
     wallet_instance->InitWalletFlags(wallet_creation_flags);
 
     if (!wallet_instance->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
-        auto spk_man = wallet_instance->GetOrCreateLegacyScriptPubKeyMan();
-        spk_man->SetupGeneration(false);
+        assert(false);
```
Comment (w0xlt) [inline src/wallet/wallettool.cpp:37]:
```suggestion
assert(wallet_instance->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS));
```

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (w0xlt) [reply src/wallet/scriptpubkeyman.cpp:84]: I agree.

**Review by w0xlt — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/28710/commits/be3c13a22fa8d49fe15cde941e8030e1cd2b522f with 2 non-blocking nits commented earlier.

Regarding the removed RPC functions, I think less is better in this case, and unless I'm missing something, the cases mentioned in https://github.com/bitcoin/bitcoin/issues/30175 can be performed by `importdescriptors`.

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:84]: Maybe for a followup.

File: src/wallet/wallettool.cpp (hunk 14 above)
Comment (achow101) [reply src/wallet/wallettool.cpp]: Done

**Review by w0xlt — COMMENTED** [review]
reACK https://github.com/bitcoin/bitcoin/pull/28710/commits/de054df6dc32cbd8b127c6761d9c65d95025e08f

**Comment by Sjors** [comment]
re-ACK de054df6dc32cbd8b127c6761d9c65d95025e08f

File: src/wallet/rpc/transactions.cpp [hunk 15]
```diff
@@ -946,13 +946,13 @@ RPCHelpMan rescanblockchain()
...
     return RPCHelpMan{"abortrescan",
-                "\nStops current wallet rescan triggered by an RPC call, e.g. by an importprivkey call.\n"
+                "\nStops current wallet rescan triggered by an RPC call, e.g. by an rescanblockchain call.\n"
                 "Note: Use \"getwalletinfo\" to query the scanning progress.\n",
                 {},
                 RPCResult{RPCResult::Type::BOOL, "", "Whether the abort was successful"},
                 RPCExamples{
             "\nImport a private key\n"
-            + HelpExampleCli("importprivkey", "\"mykey\"") +
+            + HelpExampleCli("rescanblockchain", "") +
```
Comment (rkrux) [inline src/wallet/rpc/transactions.cpp:955]: Besides the typo `s/an/a` above, the comment "Import a private key" doesn't go well with `rescanblockchain`.

File: src/wallet/rpc/backup.cpp [hunk 16]
```diff
@@ -406,842 +122,6 @@ RPCHelpMan removeprunedfunds()
...
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Range should not be specified for an un-ranged descriptor");
-    } else if (parsed_descs.at(0)->IsRange()) {
-        if (!data.exists("range")) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Descriptor is ranged, please specify the range");
-        }
-        std::tie(range_start, range_end) = ParseDescriptorRange(data["range"]);
-    }
-
-    // Only single key descriptors are allowed to be imported to a legacy wallet's keypool
-    bool can_keypool = parsed_descs.at(0)->IsSingleKey();
```
Comment (rkrux) [inline src/wallet/rpc/backup.cpp:1095]: In #31243, `IsSingleKey` was added with a TODO to remove it after removing legacy wallets.
I checked that no other usage of `IsSingleKey` is left, it can be removed. 

https://github.com/bitcoin/bitcoin/blob/6d5edfcc585bab3374ae14aa7918b3e178e016aa/src/script/descriptor.h#L114-L117

**Review by rkrux — COMMENTED** [review]
Concept ACK de054df6dc32cbd8b127c6761d9c65d95025e08f

Couple of points in 8ede6dea0c55bb4afefa790b83dd4da48a2f84da, can be done in follow up if no more changes being done in this PR.

**Review by pablomartin4btc — COMMENTED** [review]
post-merge ACK de054df6dc32cbd8b127c6761d9c65d95025e08f
